File: algorithms/term-3/dfs, mst/E.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.*;

public class E {
    private static int timer;
    private static ArrayList<ArrayList<Pair<Integer, Integer>>> graph;
    private static int[] times;
    private static int[] colorEdges;
    private static boolean[] usedEdges;
    private static int cntVertex;
    private static int cntEdges;
    private static Stack<Integer> stack;
    private static int maxColor;
    private static Map<Integer, Integer> multiply;

    public static void main(String[] args) {
        graphInit();
        solve();
        print();
    }

    private static void graphInit() {

        timer = 0;
        maxColor = 0;
        graph = new ArrayList<>();

        stack = new Stack<>();
        multiply = new LinkedHashMap<>();
        Map<Pair<Integer, Integer>, Integer> edgesId = new LinkedHashMap<>();
        InputReader scanner = new InputReader(System.in);
//        Scanner scanner = new Scanner(System.in);
        cntVertex = scanner.nextInt();
        cntEdges = scanner.nextInt();

        times = new int[cntVertex + 1];
        colorEdges = new int[cntEdges + 1];
        usedEdges = new boolean[cntEdges + 1];

        for (int i = 0; i < cntVertex; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < cntEdges; i++) {
            int u = scanner.nextInt() - 1;
            int v = scanner.nextInt() - 1;
            if (!edgesId.containsKey(new Pair<>(u, v))) {
                edgesId.put(new Pair<>(u, v), i);
                edgesId.put(new Pair<>(v, u), i);
            } else {
                multiply.put(i, edgesId.get(new Pair<>(v, u)));
            }
            graph.get(u).add(new Pair<>(v, i));
            graph.get(v).add(new Pair<>(u, i));
        }
    }

    private static void solve() {
        for (int i = 0; i < cntVertex; i++) {
            if (times[i] == 0) {
                dfs(i, -1);
            }
        }
    }

    private static int dfs(int v, int parent) {
        timer++;
        times[v] = timer;
        int minTime = timer;
        for (int i = 0; i < graph.get(v).size(); i++) {
            int u = graph.get(v).get(i).getFirst();
            int id = graph.get(v).get(i).getSecond();

            if (u != parent) {
                int t, currentSize = stack.size();
                if (!usedEdges[id]) {
                    stack.add(id);
                    usedEdges[id] = true;
                }
                if (times[u] == 0) {
                    t = dfs(u, v);
                    if (t >= times[v]) {
                        maxColor++;
                        while (stack.size() != currentSize) {
                            int edge = stack.peek();
                            colorEdges[edge] = maxColor;
                            stack.pop();
                        }
                    }
                } else {
                    t = times[u];
                }
                minTime = Math.min(minTime, t);
            }
        }
        return minTime;
    }

    private static void print() {
        System.out.println(maxColor);
        if (multiply.size() != 0) {
            for (int i = 0; i < cntEdges; i++) {
                int curr = colorEdges[i];
                if (multiply.containsKey(i)) {
                    curr = colorEdges[multiply.get(i)];
                }
                System.out.print(curr + " ");
            }
        } else {
            for (int i = 0; i < cntEdges; i++) {
                System.out.print(colorEdges[i] + " ");
            }
        }
    }

    private static class Pair<T, V> {
        private final T first;
        private final V second;

        Pair(T first, V second) {
            this.first = first;
            this.second = second;
        }

        T getFirst() {
            return first;
        }

        V getSecond() {
            return second;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Pair<?, ?> pair = (Pair<?, ?>) o;
            return Objects.equals(first, pair.first) &&
                    Objects.equals(second, pair.second);
        }

        @Override
        public int hashCode() {
            return Objects.hash(first, second);
        }
    }

    private static class InputReader {
        BufferedReader reader;
        StringTokenizer tokenizer;

        InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }

        String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
            return tokenizer.nextToken();
        }

        public int nextInt() {
            return Integer.parseInt(next());
        }

    }
}

File: algorithms/term-3/dfs, mst/E.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <map>

using namespace std;

int timer, cnt_vertex, cnt_edges, max_color;
vector<vector<pair<int, int>>> graph;
vector<int> times;
vector<int> color_edges;
stack<int> st{};
vector<bool> used_edges;
map<int, int> multiply{};

void graph_init() {
    cin >> cnt_vertex >> cnt_edges;
    timer = 0;
    max_color = 0;
    graph.resize(cnt_vertex);
    times.resize(cnt_vertex, 0);
    map<pair<int, int>, int> edges_id{};
    color_edges.resize(cnt_edges, -1);
    used_edges.resize(cnt_edges, false);

    for (int i = 0; i < cnt_edges; i++) {
        int u, v;
        cin >> u >> v;
        u--;
        v--;
        if (edges_id.find(make_pair(u, v)) == edges_id.end()) {
            edges_id[make_pair(u, v)] = i;
            edges_id[make_pair(v, u)] = i;
        } else {
            multiply[i] = edges_id[make_pair(v, u)];
        }
        graph[u].push_back(make_pair(v, i));
        graph[v].push_back(make_pair(u, i));
    }
}

int dfs(int v, int parent) {
    timer++;
    times[v] = timer;
    int min_time = timer;
    for (unsigned int i = 0; i < graph[v].size(); i++) {
        int u = graph[v][i].first, id = graph[v][i].second;
        if (u != parent) {
            int t, current_size = st.size();
            if (!used_edges[id]) {
                st.push(id);
                used_edges[id] = true;
            }
            if (times[u] == 0) {
                t = dfs(u, v);
                if (t >= times[v]) {
                    max_color++;
                    while (st.size() != current_size) {
                        int edge = st.top();
                        color_edges[edge] = max_color;
                        st.pop();
                    }
                }
            } else {
                t = times[u];
            }
            min_time = min(min_time, t);
        }
    }
    return min_time;
}

void solve() {
    for (int i = 0; i < cnt_vertex; i++) {
        if (times[i] == 0) {
            dfs(i, -1);
        }
    }
}

void print() {
    cout << max_color << '\n';
    for (int i = 0; i < cnt_edges; i++) {
        int curr = color_edges[i];
        if (multiply.find(i) != multiply.end()) {
            curr = color_edges[multiply[i]];
        }
        cout << curr << ' ';
    }
}

int main() {

    graph_init();
    solve();
    print();
}